### PR TITLE
fix(ci): Use `pytest --ignore` to Skip Tests Based on Framework

### PR DESCRIPTION
**Pull request by whoisj**
## Overview:
   
This change uses the `inputs.framework` value to determine if specific tests should be skipped.
This is useful as specific tests will automatically fail due to dependencies missing in the containers for specific frameworks.
    
Thanks to @kthui for the recommendation.

## Details:

Removes previous explicit import failure => ignore logic which could hide actual errors in non-Triton frameworks.

**🚫 This PR is NOT linked to an issue**:
- [x] Confirmed — no related issue

## Summary by CodeRabbit

- **Tests**
  - Updated continuous integration test selection to apply framework-specific exclusions while retaining tracing options.
  - Adjusted fault-tolerance and serving test suites so they run without automatically skipping entire modules when PyTorch is unavailable.

### CODE REVIEW

File: .github/workflows/shared-test.yml [hunk 1]
```diff
@@ -193,7 +193,14 @@ jobs:
         DD_CIVISIBILITY_ENABLED: 'true'
         DD_CIVISIBILITY_AGENTLESS_ENABLED: 'true'
         DD_CIVISIBILITY_AUTO_INSTRUMENTATION_PROVIDER: github
-        PYTEST_ADDOPTS: --ddtrace
+        PYTEST_ADDOPTS: >
+          --ddtrace
+          ${{ inputs.framework == 'triton' && '--ignore=tests/fault_tolerance/test_canary_rank_pause.py' || '' }}
```
Comment (dmitry-tokarev-nv) [inline .github/workflows/shared-test.yml:198]: **P3 — `framework` is now a test-selection input, but its description still calls it a label, and omits `triton`.**

Before this change `inputs.framework` only tagged metadata. It now decides which test files are collected, while its description at line 14 still reads *"labels test metadata and coverage/Allure artifacts"*, and the value list `(vllm, sglang, trtllm, dynamo)` does not include `triton` — the one value every expression here branches on.

`framework` is free-form caller-supplied text with no validation, and `!= 'triton'` is the default branch: anything that is not exactly `triton` drops `components/src/dynamo/triton/tests/` and `tests/frontend/grpc/test_triton_identity.py`. I checked how loud that would be before rating it — every `pytest.mark.triton` in the repo lives inside those two paths (10 files), so a wrong value on the Triton lane leaves `-m "pre_merge and triton and gpu_0"` selecting nothing; measured on pytest 9.0.3 that exits **5**, and `exit ${TEST_EXIT_CODE}` in `pytest-local` fails the step. It breaks loudly rather than silently, which is why this is P3 and not higher. Still worth two words in the description so the next person editing a `framework:` value for labelling reasons knows it moves tests:

```diff
       framework:
-        description: 'Backend the test image was built with (vllm, sglang, trtllm, dynamo); labels test metadata and coverage/Allure artifacts'
+        description: 'Backend the test image was built with (vllm, sglang, trtllm, triton, dynamo); labels test metadata and coverage/Allure artifacts, and selects the framework-specific pytest --ignore list below'
```

(Offered as a diff rather than a `suggestion` block because line 14 is outside this PR's hunk, so a suggestion could not anchor there.)

File: .github/workflows/shared-test.yml [hunk 2]
```diff
@@ -193,7 +193,14 @@ jobs:
...
+          ${{ inputs.framework == 'triton' && '--ignore=tests/serve/test_vllm.py' || '' }}
+          ${{ inputs.framework == 'triton' && '--ignore=tests/serve/test_sglang.py' || '' }}
+          ${{ inputs.framework == 'triton' && '--ignore=tests/serve/test_trtllm.py' || '' }}
```
Comment (dmitry-tokarev-nv) [inline .github/workflows/shared-test.yml:201]: **P3 — the "needs torch" fact now lives only in this workflow, so the documented in-container repro breaks.**

The guards this PR deletes were a property of the test files, so they held for every pytest invocation. The replacement holds only where this `PYTEST_ADDOPTS` is in the environment. `.ai/pytest-guidelines.md` tells people to reproduce a CI failure by pulling the image CI built and running `python3 -m pytest` inside it ("Pull the container, reproduce, fix, verify — in that order"). On the Triton image that flow now aborts before running anything.

Measured on pytest 9.0.3, four modules unguarded exactly as at head, no `--ignore`:

```
ERROR tests/fault_tolerance/test_canary_rank_pause.py
ERROR tests/serve/test_sglang.py
ERROR tests/serve/test_trtllm.py
ERROR tests/serve/test_vllm.py
!!!!!!!!!!!!!!!!!!! Interrupted: 4 errors during collection !!!!!!!!!!!!!!!!!!!!
```

With these four `--ignore`s applied the identical command collects cleanly and exits 0, and with `--continue-on-collection-errors` (what CI passes) the unguarded run exits 1 rather than being interrupted. CI itself is fine — verified in the review body — so this is a developer-ergonomics cost, not a correctness one.

Two cheap mitigations, either is enough. A comment here naming *why* these four are excluded: the rationale the deleted guards carried (`dynamo.common.multimodal`'s package `__init__` imports torch eagerly, and the serve modules reach it) now exists nowhere in the tree, so a reader of this list sees four paths and no reason. And/or a line in `.ai/pytest-guidelines.md` for the Triton image.

Related, worth knowing when maintaining this list: a later rename of any of these four files makes its `--ignore` match nothing with no diagnostic at all — I confirmed pytest neither warns nor fails on a stale `--ignore` — and the symptom then surfaces as a collection error on the Triton lane rather than as a bad path here.

File: .github/workflows/shared-test.yml [hunk 3]
```diff
@@ -193,7 +193,14 @@ jobs:
...
+          ${{ inputs.framework != 'triton' && '--ignore=tests/frontend/grpc/test_triton_identity.py' || '' }}
+          ${{ inputs.framework != 'triton' && '--ignore=components/src/dynamo/triton/tests/' || '' }}
```
Comment (dmitry-tokarev-nv) [inline .github/workflows/shared-test.yml]: **P3 — these two exclusions duplicate a mechanism that already covers them.**

The PR description motivates the change with tests that "will automatically fail due to dependencies missing in the containers". That argument covers lines 198–201, but not these two: on a non-Triton image these paths do not fail, because they are already excluded.

- `components/src/dynamo/triton/tests/conftest.py` defines `pytest_ignore_collect` and is registered as a *global* plugin via the `pytest11` entry point in `pyproject.toml` (`triton_tests = "dynamo.triton.tests.conftest"`), so it ignores every `test_triton_*.py` in the repo whenever `tritonclient`/`tritonserver` cannot be imported.
- `tests/frontend/grpc/test_triton_identity.py` additionally opens with `pytest.importorskip("tritonserver")`.

Measured: on both `dynamo`-framework CPU lanes — the only non-Triton lanes whose markers can select `triton`-marked tests — neither path appears in the executed node-ID set at base or at head, so these two lines change nothing today.

The cost is two independent sources of truth for the same exclusion, keyed differently: file-name plus dependency probe on one side, path plus the `framework` input on the other. They can drift — if a non-Triton image ever ships the Triton client, the conftest hook stops excluding and this line keeps excluding. Either drop them, or say in a comment that they are a belt-and-braces guard, so the next reader does not take them for the load-bearing mechanism.

**Review by dmitry-tokarev-nv — APPROVED** [review]
**Intent, from the diff and the PR body:** move the "this file cannot be imported on the Triton image" decision out of four test modules and into `shared-test.yml`, so a genuine `ModuleNotFoundError` in those files stops being swallowed by a `try: import torch` guard.

`pytest --ignore` is a collection-time exclusion that leaves no trace in the result count, so I established the executed set on both sides rather than reading it off the diff.

### No test silently stopped running

**Triton lane** (`framework == 'triton'`; `pr.yaml:936` and `post-merge-ci.yml:639` are its only callers, both through this workflow, so nothing runs the Triton image outside this `PYTEST_ADDOPTS`):

| | passed | skipped | total |
|---|---|---|---|
| base — main `165fca70`, post-merge `triton-runtime / CPU Test` | 117 | 28 | 145 |
| head — this PR, pre-merge `triton-runtime / CPU Test` | 117 | 24 | 141 |

Same 117 passes. The 4-record delta is exactly the four module-level skips this PR deletes — base logs `SKIPPED [1] tests/serve/test_vllm.py:19: torch not available in this image` and its three siblings, head logs none of them. From the `-v` output the lane collects exactly the eight `components/src/dynamo/triton/tests/test_triton_*.py` files plus `tests/frontend/grpc/test_triton_identity.py`, on both sides.

**Non-Triton lanes.** The `dynamo`-framework CPU jobs are the only ones whose marker expression (`... and not (vllm or sglang or trtllm) ...`) can select `triton`-marked tests. Base collected 5751, head 5699. Rather than trust the counts I diffed the executed node IDs: all 52 are in `components/src/dynamo/experimental/tests/test_llm.py` (18), `components/src/dynamo/frontend/tests/test_jemalloc.py` (13), `components/src/dynamo/experimental/tests/test_endpoint.py` (9), `lib/bindings/python/tests/*` (11) and `benchmarks/router/tests/test_aiperf_commands.py` (2) — tests main gained in the 37 commits after this PR's merge base, not anything the new `--ignore`s removed. Neither newly-ignored path appears in the executed set on either side. Collection errors: 0 on both. `planner` also passes `framework: dynamo`, but its image's `/workspace` has no `tests/serve/` at all, so it is untouched either way.

### The mechanics hold

Parsing the folded scalar and substituting the expressions yields `--ddtrace --ignore=… --ignore=…\n` with runs of blanks; pytest `shlex.split`s `PYTEST_ADDOPTS`, so the blanks and the trailing newline disappear. On pytest 9.0.3 (the version in the CI image), from the cwd the action actually uses (`cd /workspace`, `rootdir: /workspace`), against a tree mirroring these paths: all six relative paths resolve, and `--ignore=components/src/dynamo/triton/tests/` does exclude the directory despite the trailing slash. I also confirmed the failure mode this class of change usually hides — `--ignore` on a non-existent path is completely silent, no warning, exit 0 — and checked all six against the tree at head; none is stale.

`tests/fault_tolerance/test_canary_rank_pause.py` is the only importer of the three now-unguarded `tests/serve/test_*` modules, and it is itself in the ignore list, so the set is closed.

Three P3 notes inline, all non-blocking.


---

**Approved.** Zero P0/P1; three P3 notes above, none blocking. Deliberately deferred: the `--ignore` list is maintained by hand and will need re-checking if any of those four files moves — noted inline rather than held against this PR. I authored no commits on this PR.

File: .github/workflows/shared-test.yml (hunk 3 above)
Comment (whoisj) [reply .github/workflows/shared-test.yml]: removed the trailing two `framework != triton` lines.

**Review by kthui — APPROVED** [review]
